File: src/omie_mcp/oauth.py
"""Authorization server OAuth 2.1 mínimo para o bridge HTTP.

Clientes MCP remotos (claude.ai entre eles) autenticam por OAuth, não por key na URL.
Este módulo transforma a tela de key em um `authorization_endpoint`: cada usuário digita
a key no navegador e recebe um access token pessoal. A key nunca circula na URL do
conector, então a mesma URL pode ser distribuída para a organização inteira e só quem
souber a key consegue concluir o vínculo.

Tudo é stateless — códigos e tokens são assinados com HMAC derivado da MCP_BRIDGE_KEY,
então sobrevivem a restart e a múltiplas instâncias do App Service sem storage externo.
Trocar a MCP_BRIDGE_KEY invalida todos os tokens já emitidos, o que é o comportamento
desejado para revogar acesso.

Referências: OAuth 2.1, RFC 7591 (registro dinâmico), RFC 8414 (metadata do AS),
RFC 8707 (resource indicators), RFC 9207 (parâmetro iss) e RFC 9728 (metadata do
recurso protegido).
"""

import hashlib
import hmac
import html
import json
import os
import secrets
import time
from base64 import urlsafe_b64decode, urlsafe_b64encode
from urllib.parse import urlencode, urlsplit

SCOPE = "omie"
ACCESS_TOKEN_TTL = 3600
REFRESH_TOKEN_TTL = 90 * 24 * 3600
AUTH_CODE_TTL = 300

# Para onde o authorization_endpoint aceita redirecionar depois do vínculo. Restringir
# isso é o que impede o endpoint de ser usado como open redirector.
DEFAULT_REDIRECT_HOSTS = (
    "claude.ai",
    "www.claude.ai",
    "claude.com",
    "www.claude.com",
    "localhost",
    "127.0.0.1",
)

LOGIN_PAGE = """<!doctype html>
<html lang="pt-BR">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>omie-mcp bridge</title>
    <style>
        :root { color-scheme: dark; --bg: #0b1220; --panel: #111a2e; --text: #e5eefb; --muted: #92a4c3; --accent: #7cc4ff; }
        * { box-sizing: border-box; }
        body {
            margin: 0; min-height: 100vh; display: grid; place-items: center;
            background: radial-gradient(circle at top, #1a2a49 0%, var(--bg) 60%);
            color: var(--text); font-family: Arial, Helvetica, sans-serif; padding: 24px;
        }
        .card {
            width: 100%; max-width: 520px; background: rgba(17, 26, 46, 0.92);
            border: 1px solid rgba(124, 196, 255, 0.22); border-radius: 20px;
            padding: 28px; box-shadow: 0 24px 80px rgba(0, 0, 0, 0.35);
        }
        h1 { margin: 0 0 8px; font-size: 28px; }
        p { margin: 0 0 18px; color: var(--muted); line-height: 1.5; }
        label { display: block; margin: 18px 0 8px; font-weight: 600; }
        input[type=password] {
            width: 100%; border: 1px solid rgba(146, 164, 195, 0.35); border-radius: 12px;
            background: #0b1020; color: var(--text); padding: 14px 16px; font-size: 16px;
            outline: none;
        }
        input[type=password]:focus { border-color: var(--accent); box-shadow: 0 0 0 3px rgba(124, 196, 255, 0.15); }
        button {
            margin-top: 16px; width: 100%; border: 0; border-radius: 12px; padding: 14px 16px;
            background: linear-gradient(135deg, #7cc4ff, #4f8cff); color: #07111f;
            font-weight: 700; font-size: 16px; cursor: pointer;
        }
        .error {
            margin: 0 0 4px; padding: 12px 14px; border-radius: 12px;
            background: rgba(255, 108, 108, 0.12); border: 1px solid rgba(255, 108, 108, 0.4);
            color: #ffb3b3; font-size: 14px;
        }
        .client { margin-top: 18px; font-size: 13px; color: var(--muted); }
    </style>
</head>
<body>
    <main class="card">
        <h1>omie-mcp bridge</h1>
        <p>__INTRO__</p>
        __ERROR__
        <form method="post" action="__ACTION__">
            <label for="key">Chave de acesso</label>
            <input id="key" name="key" type="password" placeholder="Informe a chave do bridge" autocomplete="off" autofocus required>
            __HIDDEN__
            <button type="submit">Vincular</button>
        </form>
        __CLIENT__
    </main>
</body>
</html>"""


def _b64u(raw: bytes) -> str:
    return urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def _b64u_decode(text: str) -> bytes:
    padding = "=" * (-len(text) % 4)
    return urlsafe_b64decode(text + padding)


def _now() -> int:
    return int(time.time())


def derive_secret(bridge_key: str) -> bytes:
    """Deriva a chave de assinatura. MCP_TOKEN_SECRET permite rotacionar tokens sem
    trocar a key que os usuários digitam (e vice-versa)."""
    extra = os.getenv("MCP_TOKEN_SECRET", "")
    return hashlib.sha256(f"omie-mcp\0{bridge_key}\0{extra}".encode("utf-8")).digest()


def _sign(payload: dict, secret: bytes) -> str:
    body = _b64u(json.dumps(payload, separators=(",", ":"), sort_keys=True).encode("utf-8"))
    signature = _b64u(hmac.new(secret, body.encode("ascii"), hashlib.sha256).digest())
    return f"{body}.{signature}"


def _unsign(token: str, secret: bytes, expected_typ: str) -> dict | None:
    """Valida assinatura, tipo e expiração. Devolve o payload ou None."""
    body, _, signature = token.partition(".")
    if not body or not signature:
        return None

    expected = _b64u(hmac.new(secret, body.encode("ascii"), hashlib.sha256).digest())
    if not hmac.compare_digest(signature, expected):
        return None

    try:
        payload = json.loads(_b64u_decode(body))
    except (ValueError, TypeError):
        return None

    if not isinstance(payload, dict) or payload.get("typ") != expected_typ:
        return None
    if payload.get("exp", 0) <= _now():
        return None
    return payload


def allowed_redirect_hosts() -> frozenset[str]:
    configured = os.getenv("MCP_ALLOWED_REDIRECT_HOSTS", "")
    if not configured.strip():
        return frozenset(DEFAULT_REDIRECT_HOSTS)
    return frozenset(h.strip().lower() for h in configured.split(",") if h.strip())


class OAuthProvider:
    """Emite e valida credenciais para o bridge. Sem estado em memória."""

    def __init__(self, bridge_key: str) -> None:
        self.bridge_key = bridge_key
        self.secret = derive_secret(bridge_key)
        self.allowed_hosts = allowed_redirect_hosts()

    # ------------------------------------------------------------------ metadata

    def protected_resource_metadata(self, base_url: str, resource_path: str) -> dict:
        return {
            "resource": f"{base_url}{resource_path}",
            "authorization_servers": [base_url],
            "scopes_supported": [SCOPE],
            "bearer_methods_supported": ["header"],
        }

    def authorization_server_metadata(self, base_url: str) -> dict:
        return {
            "issuer": base_url,
            "authorization_endpoint": f"{base_url}/authorize",
            "token_endpoint": f"{base_url}/token",
            "registration_endpoint": f"{base_url}/register",
            "scopes_supported": [SCOPE],
            "response_types_supported": ["code"],
            "response_modes_supported": ["query"],
            "grant_types_supported": ["authorization_code", "refresh_token"],
            "token_endpoint_auth_methods_supported": ["none"],
            "code_challenge_methods_supported": ["S256"],
            "authorization_response_iss_parameter_supported": True,
        }

    # ------------------------------------------------------- registro dinâmico

    def register_client(self, request: dict) -> dict:
        """RFC 7591. O client_id é um blob assinado com os redirect_uris dentro, então
        não precisamos guardar registro nenhum para validar o redirect depois."""
        redirect_uris = request.get("redirect_uris")
        if not isinstance(redirect_uris, list) or not redirect_uris:
            raise OAuthError("invalid_redirect_uri", "redirect_uris é obrigatório")

        for uri in redirect_uris:
            if not isinstance(uri, str) or not self.is_redirect_allowed(uri):
                raise OAuthError("invalid_redirect_uri", f"redirect_uri não permitido: {uri}")

        issued_at = _now()
        client_id = _sign(
            {
                "typ": "client",
                "ru": redirect_uris,
                "iat": issued_at,
                # O client_id não expira na prática, mas _unsign exige `exp`.
                "exp": issued_at + 10 * 365 * 24 * 3600,
            },
            self.secret,
        )

        response = {
            "client_id": client_id,
            "client_id_issued_at": issued_at,
            "redirect_uris": redirect_uris,
            "grant_types": ["authorization_code", "refresh_token"],
            "response_types": ["code"],
            "token_endpoint_auth_method": "none",
            "scope": SCOPE,
        }
        if isinstance(request.get("client_name"), str):
            response["client_name"] = request["client_name"]
        return response

    # -------------------------------------------------------------- redirect_uri

    def is_redirect_allowed(self, redirect_uri: str) -> bool:
        parts = urlsplit(redirect_uri)
        if parts.fragment:
            return False
        if parts.scheme == "https":
            return (parts.hostname or "").lower() in self.allowed_hosts
        # http só para desenvolvimento local
        if parts.scheme == "http":
            return (parts.hostname or "").lower() in {"localhost", "127.0.0.1"}
        return False

    def redirect_uri_matches_client(self, client_id: str, redirect_uri: str) -> bool:
        """Se o client_id foi emitido por nós, o redirect tem que estar entre os
        registrados. client_id de outra origem cai só no allowlist de hosts."""
        payload = _unsign(client_id, self.secret, "client")
        if payload is None:
            return True
        return redirect_uri in payload.get("ru", [])

    # -------------------------------------------------------------------- código

    def issue_code(self, *, client_id: str, redirect_uri: str, code_challenge: str, resource: str, scope: str) -> str:
        return _sign(
            {
                "typ": "code",
                "cid": client_id,
                "ru": redirect_uri,
                "cc": code_challenge,
                "aud": resource,
                "scope": scope,
                "jti": secrets.token_urlsafe(8),
                "exp": _now() + AUTH_CODE_TTL,
            },
            self.secret,
        )

    def redeem_code(self, *, code: str, code_verifier: str, redirect_uri: str) -> dict:
        payload = _unsign(code, self.secret, "code")
        if payload is None:
            raise OAuthError("invalid_grant", "código inválido ou expirado")

        if redirect_uri and redirect_uri != payload.get("ru"):
            raise OAuthError("invalid_grant", "redirect_uri não corresponde ao do código")

        challenge = payload.get("cc") or ""
        if challenge:
            if not code_verifier:
                raise OAuthError("invalid_request", "code_verifier é obrigatório")
            digest = hashlib.sha256(code_verifier.encode("ascii")).digest()
            if not hmac.compare_digest(_b64u(digest), challenge):
                raise OAuthError("invalid_grant", "code_verifier não corresponde ao code_challenge")

        return payload

    # -------------------------------------------------------------------- tokens

    def issue_tokens(self, *, audience: str, scope: str) -> dict:
        now = _now()
        access = _sign(
            {"typ": "at", "aud": audience, "scope": scope, "iat": now, "exp": now + ACCESS_TOKEN_TTL},
            self.secret,
        )
        refresh = _sign(
            {"typ": "rt", "aud": audience, "scope": scope, "iat": now, "exp": now + REFRESH_TOKEN_TTL},
            self.secret,
        )
        return {
            "access_token": access,
            "token_type": "Bearer",
            "expires_in": ACCESS_TOKEN_TTL,
            "refresh_token": refresh,
            "scope": scope,
        }

    def refresh_tokens(self, refresh_token: str) -> dict:
        payload = _unsign(refresh_token, self.secret, "rt")
        if payload is None:
            raise OAuthError("invalid_grant", "refresh_token inválido ou expirado")
        return self.issue_tokens(audience=payload.get("aud", ""), scope=payload.get("scope", SCOPE))

    def verify_access_token(self, token: str, valid_audiences: set[str]) -> bool:
        payload = _unsign(token, self.secret, "at")
        if payload is None:
            return False
        # RFC 8707: o token tem que ter sido emitido para este recurso.
        return payload.get("aud", "") in valid_audiences

    # ------------------------------------------------------------------ resposta

    def authorization_redirect(self, *, redirect_uri: str, params: dict) -> str:
        separator = "&" if urlsplit(redirect_uri).query else "?"
        return f"{redirect_uri}{separator}{urlencode(params)}"

    def check_key(self, provided: str) -> bool:
        return hmac.compare_digest(provided, self.bridge_key)


class OAuthError(Exception):
    def __init__(self, code: str, description: str, status: int = 400) -> None:
        super().__init__(description)
        self.code = code
        self.description = description
        self.status = status

    def to_dict(self) -> dict:
        return {"error": self.code, "error_description": self.description}


def render_login_page(*, action: str, hidden: dict, error: str = "", client_name: str = "") -> str:
    hidden_html = "".join(
        f'<input type="hidden" name="{html.escape(name, quote=True)}" value="{html.escape(value, quote=True)}">'
        for name, value in hidden.items()
        if value
    )
    error_html = f'<div class="error">{html.escape(error)}</div>' if error else ""
    client_html = (
        f'<div class="client">Solicitado por <strong>{html.escape(client_name)}</strong>.</div>' if client_name else ""
    )
    intro = (
        "Informe a chave de acesso para liberar este cliente. Cada pessoa faz esse "
        "vínculo individualmente — a chave não fica salva na URL do conector."
    )
    return (
        LOGIN_PAGE.replace("__ACTION__", html.escape(action, quote=True))
        .replace("__INTRO__", intro)
        .replace("__ERROR__", error_html)
        .replace("__HIDDEN__", hidden_html)
        .replace("__CLIENT__", client_html)
    )

File: omie/src/oauth.ts
/**
 * Minimal OAuth 2.1 authorization server for the HTTP bridge.
 *
 * Remote MCP clients (claude.ai included) need an authorization server that
 * supports Dynamic Client Registration, which Microsoft Entra ID does not. So
 * the bridge is its own authorization server facing Claude, and delegates the
 * actual login to Entra (see entra.ts): /authorize sends the browser to the
 * Microsoft SSO screen, and only members of the allowed group get a code back.
 *
 * Everything is stateless — state, codes and tokens are AES-256-GCM sealed
 * with a key derived from MCP_TOKEN_SECRET (or AZURE_CLIENT_SECRET), so they
 * survive restarts and multiple App Service instances without storage. The
 * Entra refresh token travels sealed inside ours, which lets every refresh
 * re-check the user with Entra: removing someone from the group cuts access
 * within one access-token lifetime. Rotating the secret revokes everything.
 *
 * References: OAuth 2.1, RFC 7591 (dynamic client registration), RFC 8414
 * (AS metadata), RFC 8707 (resource indicators), RFC 9207 (iss parameter),
 * RFC 9728 (protected resource metadata).
 */

import { createCipheriv, createDecipheriv, createHash, randomBytes, timingSafeEqual } from "node:crypto";

export const SCOPE = "omie";
export const ACCESS_TOKEN_TTL = 3600;
export const REFRESH_TOKEN_TTL = 90 * 24 * 3600;
export const AUTH_CODE_TTL = 300;
export const LOGIN_STATE_TTL = 600;

/** Hosts the authorization endpoint may redirect back to after the link.
 *  Restricting this is what keeps /authorize from being an open redirector. */
const DEFAULT_REDIRECT_HOSTS = [
  "claude.ai",
  "www.claude.ai",
  "claude.com",
  "www.claude.com",
  "localhost",
  "127.0.0.1",
];

export class OAuthError extends Error {
  constructor(
    public code: string,
    public description: string,
    public status: number = 400,
  ) {
    super(description);
  }

  toJSON(): Record<string, string> {
    return { error: this.code, error_description: this.description };
  }
}

/** Who completed the Microsoft login. */
export interface Identity {
  sub: string;
  email: string;
  name: string;
}

function b64u(raw: Buffer): string {
  return raw.toString("base64url");
}

function now(): number {
  return Math.floor(Date.now() / 1000);
}

export function deriveKey(secret: string): Buffer {
  return createHash("sha256").update(`omie-mcp\0${secret}`, "utf8").digest();
}

/** AES-256-GCM: authenticated and confidential, since codes and refresh
 *  tokens carry the user's Entra refresh token. */
function seal(payload: Record<string, unknown>, key: Buffer): string {
  const iv = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", key, iv);
  const body = Buffer.concat([cipher.update(JSON.stringify(payload), "utf8"), cipher.final()]);
  return b64u(Buffer.concat([iv, cipher.getAuthTag(), body]));
}

function unseal(token: string, key: Buffer, expectedTyp: string): Record<string, unknown> | null {
  let raw: Buffer;
  try {
    raw = Buffer.from(token, "base64url");
  } catch {
    return null;
  }
  if (raw.length < 29) return null;

  let payload: unknown;
  try {
    const decipher = createDecipheriv("aes-256-gcm", key, raw.subarray(0, 12));
    decipher.setAuthTag(raw.subarray(12, 28));
    const plain = Buffer.concat([decipher.update(raw.subarray(28)), decipher.final()]);
    payload = JSON.parse(plain.toString("utf8"));
  } catch {
    return null;
  }
  if (typeof payload !== "object" || payload === null) return null;
  const record = payload as Record<string, unknown>;
  if (record.typ !== expectedTyp) return null;
  if (typeof record.exp !== "number" || record.exp <= now()) return null;
  return record;
}

function str(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function identityOf(payload: Record<string, unknown>): Identity {
  return { sub: str(payload.sub), email: str(payload.email), name: str(payload.name) };
}

function allowedRedirectHosts(): Set<string> {
  const configured = (process.env.MCP_ALLOWED_REDIRECT_HOSTS || "").trim();
  if (!configured) return new Set(DEFAULT_REDIRECT_HOSTS);
  return new Set(
    configured
      .split(",")
      .map((h) => h.trim().toLowerCase())
      .filter(Boolean),
  );
}

export interface TokenSet {
  access_token: string;
  token_type: "Bearer";
  expires_in: number;
  refresh_token: string;
  scope: string;
}

/** The /authorize request Claude made, parked while the user is at Microsoft. */
export interface PendingAuthorization {
  clientId: string;
  redirectUri: string;
  state: string;
  codeChallenge: string;
  resource: string;
  scope: string;
}

export interface LoginState {
  pending: PendingAuthorization;
  /** PKCE verifier and nonce of the leg between the bridge and Entra. */
  entraVerifier: string;
  nonce: string;
  /** Hash of the browser-binding cookie, so a state can't be replayed elsewhere. */
  browser: string;
}

/** Issues and validates bridge credentials. No in-memory state. */
export class OAuthProvider {
  private key: Buffer;
  private allowedHosts: Set<string>;

  constructor(secret: string) {
    this.key = deriveKey(secret);
    this.allowedHosts = allowedRedirectHosts();
  }

  // ------------------------------------------------------------- metadata

  protectedResourceMetadata(baseUrl: string, resourcePath: string): Record<string, unknown> {
    return {
      resource: `${baseUrl}${resourcePath}`,
      authorization_servers: [baseUrl],
      scopes_supported: [SCOPE],
      bearer_methods_supported: ["header"],
    };
  }

  authorizationServerMetadata(baseUrl: string): Record<string, unknown> {
    return {
      issuer: baseUrl,
      authorization_endpoint: `${baseUrl}/authorize`,
      token_endpoint: `${baseUrl}/token`,
      registration_endpoint: `${baseUrl}/register`,
      scopes_supported: [SCOPE],
      response_types_supported: ["code"],
      response_modes_supported: ["query"],
      grant_types_supported: ["authorization_code", "refresh_token"],
      token_endpoint_auth_methods_supported: ["none"],
      code_challenge_methods_supported: ["S256"],
      authorization_response_iss_parameter_supported: true,
    };
  }

  // ------------------------------------------- dynamic client registration

  /** RFC 7591. The client_id is a sealed blob embedding the redirect_uris, so
   *  no registration storage is needed to validate the redirect later. */
  registerClient(request: Record<string, unknown>): Record<string, unknown> {
    const redirectUris = request.redirect_uris;
    if (!Array.isArray(redirectUris) || redirectUris.length === 0) {
      throw new OAuthError("invalid_redirect_uri", "redirect_uris is required");
    }
    for (const uri of redirectUris) {
      if (typeof uri !== "string" || !this.isRedirectAllowed(uri)) {
        throw new OAuthError("invalid_redirect_uri", `redirect_uri not allowed: ${uri}`);
      }
    }

    const issuedAt = now();
    const clientId = seal(
      {
        typ: "client",
        ru: redirectUris,
        iat: issuedAt,
        // client_id never really expires, but unseal() requires exp.
        exp: issuedAt + 10 * 365 * 24 * 3600,
      },
      this.key,
    );

    const response: Record<string, unknown> = {
      client_id: clientId,
      client_id_issued_at: issuedAt,
      redirect_uris: redirectUris,
      grant_types: ["authorization_code", "refresh_token"],
      response_types: ["code"],
      token_endpoint_auth_method: "none",
      scope: SCOPE,
    };
    if (typeof request.client_name === "string") response.client_name = request.client_name;
    return response;
  }

  // ---------------------------------------------------------- redirect_uri

  isRedirectAllowed(redirectUri: string): boolean {
    let parsed: URL;
    try {
      parsed = new URL(redirectUri);
    } catch {
      return false;
    }
    if (parsed.hash) return false;
    const host = parsed.hostname.toLowerCase();
    if (parsed.protocol === "https:") return this.allowedHosts.has(host);
    // http only for local development
    if (parsed.protocol === "http:") return host === "localhost" || host === "127.0.0.1";
    return false;
  }

  /** If we issued the client_id, the redirect must be among the registered
   *  ones. Foreign client_ids fall back to the host allowlist alone. */
  redirectUriMatchesClient(clientId: string, redirectUri: string): boolean {
    const payload = unseal(clientId, this.key, "client");
    if (payload === null) return true;
    const registered = payload.ru;
    return Array.isArray(registered) && registered.includes(redirectUri);
  }

  // ------------------------------------------------------------ login state

  /** The `state` sent to Entra. Sealed because it carries the PKCE verifier. */
  issueLoginState(state: LoginState): string {
    return seal({ typ: "login", ...state, exp: now() + LOGIN_STATE_TTL }, this.key);
  }

  openLoginState(token: string): LoginState | null {
    const payload = unseal(token, this.key, "login");
    if (payload === null) return null;
    return payload as unknown as LoginState;
  }

  // -------------------------------------------------------------- auth code

  issueCode(pending: PendingAuthorization, identity: Identity, entraRefreshToken: string): string {
    return seal(
      {
        typ: "code",
        cid: pending.clientId,
        ru: pending.redirectUri,
        cc: pending.codeChallenge,
        aud: pending.resource,
        scope: pending.scope,
        ...identity,
        ert: entraRefreshToken,
        jti: randomBytes(8).toString("base64url"),
        exp: now() + AUTH_CODE_TTL,
      },
      this.key,
    );
  }

  redeemCode(opts: { code: string; codeVerifier: string; redirectUri: string }): {
    audience: string;
    scope: string;
    identity: Identity;
    entraRefreshToken: string;
  } {
    const payload = unseal(opts.code, this.key, "code");
    if (payload === null) throw new OAuthError("invalid_grant", "invalid or expired code");

    if (opts.redirectUri && opts.redirectUri !== payload.ru) {
      throw new OAuthError("invalid_grant", "redirect_uri does not match the code");
    }

    const challenge = str(payload.cc);
    if (challenge) {
      if (!opts.codeVerifier) throw new OAuthError("invalid_request", "code_verifier is required");
      const digest = b64u(createHash("sha256").update(opts.codeVerifier, "ascii").digest());
      const a = Buffer.from(digest, "utf8");
      const b = Buffer.from(challenge, "utf8");
      if (a.length !== b.length || !timingSafeEqual(a, b)) {
        throw new OAuthError("invalid_grant", "code_verifier does not match code_challenge");
      }
    }
    return {
      audience: str(payload.aud),
      scope: str(payload.scope) || SCOPE,
      identity: identityOf(payload),
      entraRefreshToken: str(payload.ert),
    };
  }

  // ----------------------------------------------------------------- tokens

  issueTokens(audience: string, scope: string, identity: Identity, entraRefreshToken: string): TokenSet {
    const iat = now();
    const access = seal(
      { typ: "at", aud: audience, scope, ...identity, iat, exp: iat + ACCESS_TOKEN_TTL },
      this.key,
    );
    const refresh = seal(
      { typ: "rt", aud: audience, scope, ...identity, ert: entraRefreshToken, iat, exp: iat + REFRESH_TOKEN_TTL },
      this.key,
    );
    return {
      access_token: access,
      token_type: "Bearer",
      expires_in: ACCESS_TOKEN_TTL,
      refresh_token: refresh,
      scope,
    };
  }

  openRefreshToken(refreshToken: string): {
    audience: string;
    scope: string;
    identity: Identity;
    entraRefreshToken: string;
  } {
    const payload = unseal(refreshToken, this.key, "rt");
    if (payload === null) throw new OAuthError("invalid_grant", "invalid or expired refresh_token");
    return {
      audience: str(payload.aud),
      scope: str(payload.scope) || SCOPE,
      identity: identityOf(payload),
      entraRefreshToken: str(payload.ert),
    };
  }

  /** Returns who the token belongs to, or null if it isn't valid here. */
  verifyAccessToken(token: string, validAudiences: Set<string>): Identity | null {
    const payload = unseal(token, this.key, "at");
    if (payload === null) return null;
    // RFC 8707: the token must have been issued for this resource.
    if (typeof payload.aud !== "string" || !validAudiences.has(payload.aud)) return null;
    return identityOf(payload);
  }

  // --------------------------------------------------------------- helpers

  authorizationRedirect(redirectUri: string, params: Record<string, string>): string {
    const url = new URL(redirectUri);
    for (const [name, value] of Object.entries(params)) url.searchParams.set(name, value);
    return url.toString();
  }
}

// ------------------------------------------------------------------ HTML

function escapeHtml(text: string): string {
  return text
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&#39;");
}

const PAGE = `<!doctype html>
<html lang="pt-BR">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>omie-mcp</title>
    <style>
        :root { color-scheme: dark; }
        body {
            margin: 0; min-height: 100vh; display: grid; place-items: center;
            background: radial-gradient(circle at top, #1a2a49 0%, #0b1220 60%);
            color: #e5eefb; font-family: Arial, Helvetica, sans-serif; padding: 24px;
        }
        .card {
            width: 100%; max-width: 560px; background: rgba(17, 26, 46, 0.92);
            border: 1px solid rgba(124, 196, 255, 0.22); border-radius: 20px; padding: 28px;
        }
        h1 { margin: 0 0 12px; font-size: 26px; }
        p { color: #92a4c3; line-height: 1.6; }
        .error { color: #ffb3b3; }
    </style>
</head>
<body>
    <main class="card">
        <h1>__TITLE__</h1>
        __BODY__
    </main>
</body>
</html>`;

export function renderPage(title: string, paragraphs: string[], isError = false): string {
  const cls = isError ? ' class="error"' : "";
  const body = paragraphs.map((p) => `<p${cls}>${escapeHtml(p)}</p>`).join("\n        ");
  return PAGE.replace("__TITLE__", escapeHtml(title)).replace("__BODY__", body);
}

export const INFO_PAGE = renderPage("omie-mcp", [
  "Este é o endpoint MCP do servidor. Ele não é feito para ser aberto no navegador.",
  "Adicione esta URL como conector no Claude e clique em Conectar: o login é feito " +
    "com a sua conta Microsoft, e só quem está no grupo autorizado do conector consegue acessar.",
]);
